src/toil/server/wes/abstract_backend.py

```diff
@@ -0,0 +1,237 @@
+# Modified from: https://github.com/common-workflow-language/workflow-service
+import functools
+import json
+import os
+import logging
+import tempfile
+from abc import abstractmethod
+from typing import Optional, List, Dict, Any, Tuple, Callable
+
+import connexion  # type: ignore
+from werkzeug.utils import secure_filename
+
+logger = logging.getLogger(__name__)
+
+
+class VersionNotImplementedException(Exception):
+    """
+    Raised when the requested workflow version is not implemented.
+    """
+    def __init__(self,
+                 wf_type: str, version: Optional[str] = None, supported_versions: Optional[List[str]] = None) -> None:
+        if version:
+            message = ("workflow_type '{}' requires 'workflow_type_version' to be one of '{}'.  "
+                       "Got '{}' instead.".format(wf_type, str(supported_versions), version))
+        else:
+            message = f"workflow_type '{wf_type}' is not supported."
+
+        super(VersionNotImplementedException, self).__init__(message)
+
+
+class WorkflowNotFoundException(Exception):
+    """
+    Raised when the requested run ID is not found.
+    """
+    def __init__(self) -> None:
+        super(WorkflowNotFoundException, self).__init__("The requested workflow run wasn't found.")
+
+
+class WorkflowConflictException(Exception):
+    """
+    Raised when the requested workflow is not in the expected state.
+    """
+    def __init__(self, run_id: str):
+        super(WorkflowConflictException, self).__init__(f"Workflow {run_id} exists when it shouldn't.")
+
+
+class WorkflowExecutionException(Exception):
+    """
+    Raised when an internal error occurred during the execution of the workflow.
+    """
+    pass
+
+
+def handle_errors(func: Callable[..., Any]) -> Callable[..., Any]:
+    """
+    This decorator catches errors from the wrapped function and returns a JSON
+    formatted error message with the appropriate status code defined by the
+    GA4GH WES spec.
+    """
+
+    def error(msg: Any, code: int = 500) -> Tuple[Dict[str, Any], int]:
+        logger.warning(f"Exception raised when calling '{func.__name__}()':", exc_info=True)
+        return {"msg": str(msg), "status_code": code}, code
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):  # type: ignore
+        try:
+            return func(*args, **kwargs)
+        except (TypeError, ValueError) as e:
+            return error(e, code=400)  # the request is malformed
```

Comment:
Shouldn't this be a 500? Do we actually validate user requests by just forging ahead and making sure we don't call a method on an object that doesn't have it in *our* code? What if we run into these because the user request is innocuous but the server's internal data structures are in a bad state?

Reply:
Ah yes, this was originally intended to catch the `ValueError`s that may be thrown in `collect_attachments()`, but I've made a custom error for that now.